core/src/test/java/org/apache/iceberg/catalog/CatalogTests.java

```diff
@@ -516,6 +517,29 @@ public void testLoadMissingTable() {
         () -> catalog.loadTable(ident));
   }
 
+  @Test
+  public void testRenameTable() {
+    C catalog = catalog();
+
+    if (requiresNamespaceCreate()) {
+      catalog.createNamespace(NS);
+    }
+
+    Assert.assertFalse("Source table should not exist before create", catalog.tableExists(TABLE));
+
+    catalog.buildTable(TABLE, SCHEMA).create();
+    Assert.assertTrue("Table should exist after create", catalog.tableExists(TABLE));
+
+    Assert.assertFalse("Destination table should not exist before rename", catalog.tableExists(RENAMED_TABLE));
+
+    catalog.renameTable(TABLE, RENAMED_TABLE);
+    Assert.assertTrue("Table should exist with new name", catalog.tableExists(RENAMED_TABLE));
+    Assert.assertFalse("Original table should no longer exist", catalog.tableExists(TABLE));
+
+    catalog.dropTable(RENAMED_TABLE);
+    assertEmpty("Should not contain table after drop", catalog, NS);
+  }
```

Comment:
This test is pretty basic as I didn't want to make the PR too large.

Let me know if I should add more `CatalogTests` for the `renameTable` operation in this PR 🙂

Reply:
I'd prefer adding tests for missing source table, existing destination table, and others. It's a good time to add them.

Reply:
I added test for missing source and existing destination.

However, the JdbcCatalog does not throw the correct error for existing destination table, which can be verified by looking at the current JdbcCatalog tests: https://github.com/apache/iceberg/blob/62f3ca334c4bf377b1aaf9b7386bec1bfaddb5cd/core/src/test/java/org/apache/iceberg/jdbc/TestJdbcCatalog.java#L385-L390

I have put togeher a hacky fix to it in this PR, but I think we should just throw an `@Ignore` on this particular test and then fix the jdbc catalog in a separate PR. All of the inforrmation from the error was very SQLIte specific and I'm not sure of a general way to handle this. Open to suggestions.

Reply:
I hit the same problem when testing create table atomicity and [fixed it similarly](https://github.com/apache/iceberg/blob/master/core/src/main/java/org/apache/iceberg/jdbc/JdbcTableOperations.java#L126-L129). There's more background in [this thread](https://github.com/apache/iceberg/pull/4319#pullrequestreview-909504774).

I'd probably use the same logic as in that fix, which checks the message for "constraint failed".

Reply:
Updated the check.